qiskit/visualization/transition_visualization.py

```diff
@@ -0,0 +1,251 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Visualization function for animation of state transitions by applying gates to single qubit.
+"""
+from math import sin, cos, acos, sqrt
+import numpy as np
+
+try:
+    from matplotlib import pyplot as plt
+    from matplotlib import animation
+    from mpl_toolkits.mplot3d import Axes3D
+    HAS_MATPLOTLIB = True
+except ImportError:
+    HAS_MATPLOTLIB = False
+
+try:
+    from IPython.display import HTML
+    HAS_IPYTHON = True
+except ImportError:
+    HAS_IPYTHON = False
+
+from qiskit.visualization.bloch import Bloch
```

Comment:
This import needs to be moved into the the try block before `HAS_MATPLOTLIB`, or alternatively needs to be inside an `if HAS_MATPLOTLIB` block. Normally what I suggest is just use the HAS_MATPLOTLIB variable from another module, like:

```
from qiskit.visualization.matplotlib import HAS_MATPLOTLIB

if HAS_MATPLOTLIB:
    from matplotlib import pyplot as plt
    from matplotlib import animation
    from mpl_toolkits.mplot3d import Axes3D
    from qiskit.visualization.bloch import Bloch
```

The test failure you're hitting is caused by matplotlib's default backend for osx does not work on every osx installation of python, which includes the CI nodes. So when pyplot gets imported it raises an exception. This is the main reason why we have all the visualization code and dependencies as optional features of terra and why all the other matplotlib imports are in try blocks or if blocks to make sure we don't unconditionally import matplotlib at the module level.

Reply:
Thank you for your help.

I tried to use what you suggested, 
```
from qiskit.visualization.matplotlib import HAS_MATPLOTLIB

if HAS_MATPLOTLIB:
    from matplotlib import pyplot as plt
    from matplotlib import animation
    from mpl_toolkits.mplot3d import Axes3D
    from qiskit.visualization.bloch import Bloch
```

The thing is, if I import matplotlib from the package visualization, it means I won't be able to import the actual package matplotlib which contains subpackage animation. I could make it so visualization/matplotlib.py includes animation but I'm not sure if it's okay.

Also with this approach the imports from `qiskit.visualization.matplotlib import HAS_MATPLOTLIB` and from `qiskit.visualization.bloch import Bloch` makes pylint complain that they're ungrouped.

And so my code is 

```
try:
    from matplotlib import pyplot as plt
    from matplotlib import animation
    from mpl_toolkits.mplot3d import Axes3D
    from qiskit.visualization.bloch import Bloch
    from qiskit.visualization.exceptions import VisualizationError
    HAS_MATPLOTLIB = True
```

And pylint is happy. 

Is it okay or should I rather do as you suggest and import visualization/matplotlib.py and change it so that it includes animation subpackage too even if it means imports will be ungrouped and pylint will be unhappy?